src/transformers/models/cogvlm/modeling_cogvlm.py

```diff
@@ -0,0 +1,1056 @@
+                batch_size = pixel_values.shape[0]
+                vision_mask = (
+                    torch.tensor([1] * self.num_vision_tokens).repeat(batch_size, 1).to(attention_mask.device)
+                )
+                attention_mask = torch.cat(
+                    [attention_mask[:, :-1], vision_mask, attention_mask[:, -1].repeat(batch_size, 1)], dim=1
+                )
+        else:
+            # not allow for inputs_embeds, because we want to process image feature
+            # assert input_ids is not None and inputs_embeds is None, f"{input_ids} {inputs_embeds}"
+
+            if pixel_values is not None and input_ids is not None:
+                # multi-modality
+                if len(input_ids) != len(pixel_values):
+                    raise ValueError("Make sure to pass as many texts as images")
+
+                # prepend the input_ids and token_type_ids with image tokens
+                batch_size = input_ids.shape[0]
+
+                vision_input_ids = (
+                    torch.tensor([self.config.bos_token_id] + [self.config.pad_token_id] * self.num_vision_tokens)
+                    .repeat(batch_size, 1)
+                    .to(input_ids.device)
+                )
+                vision_token_type_ids = (
+                    torch.tensor([LANGUAGE_TOKEN_TYPE] + [VISION_TOKEN_TYPE] * self.num_vision_tokens)
+                    .repeat(batch_size, 1)
+                    .to(token_type_ids.device)
+                )
+
+                input_ids = torch.cat([vision_input_ids, input_ids[:, 1:]], dim=1)
+                token_type_ids = torch.cat([vision_token_type_ids, token_type_ids[:, 1:]], dim=1)
+                attention_mask = torch.tensor([1] * input_ids.shape[1]).repeat(batch_size, 1).to(input_ids.device)
+
+                inputs_embeds = self.embed_tokens(input_ids)
+                images_features = self.encode_images(pixel_values)
+                images_features = images_features.reshape(-1, images_features.shape[-1])
+                images_features = images_features.to(dtype=inputs_embeds.dtype, device=inputs_embeds.device)
+                inputs_embeds = inputs_embeds.index_put([token_type_ids == VISION_TOKEN_TYPE], images_features)
+
+            else:
+                # TODO verify single-modality
+                if token_type_ids is None:
+                    token_type_ids = (
+                        torch.ones_like(input_ids, dtype=torch.long, device=input_ids.device) * LANGUAGE_TOKEN_TYPE
+                    )
+                assert not (
+                    token_type_ids == VISION_TOKEN_TYPE
+                ).any(), f"{(token_type_ids == VISION_TOKEN_TYPE).sum()}"
+                if inputs_embeds is None:
+                    inputs_embeds = self.embed_tokens(input_ids)
+
+            if position_ids is None:
+                position_ids = build_position_ids(token_type_ids, attention_mask)
+            input_ids = None
+
+        return self.llm_forward(
```

Comment:
would be simpler to define the LLM no? (I mean a bit more clean)

Reply:
Alright let's just change the name. The forward should be the `llm_forward`, and the pre-processing should be `pre-process` or `prepare_masks` or whatever. The forward should be the part where we call the layers etx